File: languages/go/secrets.go
package sdk

type SecretsInterface interface {
	Create(options CreateSecretOptions) (*CreateSecretResponseSecret, error)
	List(options ListSecretsOptions) (*[]SecretElement, error)
	Get(options GetSecretOptions) (*GetSecretResponseSecret, error)
	Delete(options DeleteSecretOptions) (*DeleteSecretResponseSecret, error)
	Update(options UpdateSecretOptions) (*UpdateSecretResponseSecret, error)
}

type Secrets struct {
	CommandRunner CommandRunnerInterface
}

func NewSecrets(commandRunner CommandRunnerInterface) *Secrets {
	return &Secrets{CommandRunner: commandRunner}
}

func (s *Secrets) executeCommand(command Command, target interface{}) error {
	responseStr, err := s.CommandRunner.RunCommand(command)
	if err != nil {
		return err
	}
	return checkSuccessAndError(responseStr, target)
}

func (s *Secrets) Create(options CreateSecretOptions) (*CreateSecretResponseSecret, error) {
	command := Command{
		CreateSecret: &options,
	}

	var response CreateSecretResponse
	if err := s.executeCommand(command, &response); err != nil {
		return nil, err
	}

	return &response.Secret, nil
}

func (s *Secrets) List(options ListSecretsOptions) (*[]SecretElement, error) {
	command := Command{
		ListSecrets: &options,
	}

	var response ListSecretsResponse
	if err := s.executeCommand(command, &response); err != nil {
		return nil, err
	}

	return &response.Secrets, nil
}

func (s *Secrets) Get(options GetSecretOptions) (*GetSecretResponseSecret, error) {
	command := Command{
		GetSecret: &options,
	}

	var response GetSecretResponse
	if err := s.executeCommand(command, &response); err != nil {
		return nil, err
	}

	return &response.Secret, nil
}

func (s *Secrets) Delete(options DeleteSecretOptions) (*DeleteSecretResponseSecret, error) {
	command := Command{
		DeleteSecret: &options,
	}

	var response DeleteSecretResponse
	if err := s.executeCommand(command, &response); err != nil {
		return nil, err
	}

	return &response.Secret, nil
}

func (s *Secrets) Update(options UpdateSecretOptions) (*UpdateSecretResponseSecret, error) {
	command := Command{
		UpdateSecret: &options,
	}

	var response UpdateSecretResponse
	if err := s.executeCommand(command, &response); err != nil {
		return nil, err
	}

	return &response.Secret, nil
}

File: languages/ruby/infisical-sdk/lib/clients/secrets.rb
# frozen_string_literal: true

require 'json'

require_relative '../extended_schemas/schemas'

module InfisicalSDK
  # Manage Infisical secrets.
  class SecretsClient
    def initialize(command_runner)
      @command_runner = command_runner
    end

    # rubocop:disable Metrics/ParameterLists
    # rubocop:disable Metrics/MethodLength
    def get(
      secret_name:,
      project_id:,
      environment:,
      path: nil,
      include_imports: nil,
      type: nil
    )

      response = run_command(get_secret: GetSecretOptions.new(
        secret_name: secret_name,
        project_id: project_id,
        environment: environment,
        path: path,
        include_imports: include_imports,
        get_secret_options_type: type
      ))

      secrets_response = ResponseForGetSecretResponse.from_json!(response).to_dynamic
      error_handler(secrets_response)


      secrets_response['data']['secret']
    end

    def list(
      project_id:,
      environment:,
      path: nil ,
      attach_to_process_env: nil,
      expand_secret_references: nil,
      recursive: nil,
      include_imports: nil
    )
      response = run_command(list_secrets: ListSecretsOptions.new(
        project_id: project_id,
        environment: environment,
        path: path,
        include_imports: include_imports,
        recursive: recursive,
        attach_to_process_env: attach_to_process_env,
        expand_secret_references: expand_secret_references,
      ))

      secrets_response = ResponseForListSecretsResponse.from_json!(response).to_dynamic
      error_handler(secrets_response)

      secrets_response['data']['secrets']
    end

    def update(
      secret_name:,
      secret_value:,
      project_id:,
      environment:,
      path: nil,
      skip_multiline_encoding: nil,
      type: nil
    )
      response = run_command(update_secret: UpdateSecretOptions.new(
        secret_name: secret_name,
        secret_value: secret_value,
        project_id: project_id,
        environment: environment,
        path: path,
        skip_multiline_encoding: skip_multiline_encoding,
        update_secret_options_type: type
      ))

      secrets_response = ResponseForUpdateSecretResponse.from_json!(response).to_dynamic
      error_handler(secrets_response)


      secrets_response['data']['secret']
    end

    def create(
      secret_name:,
      secret_value:,
      project_id:,
      environment:,
      secret_comment: nil,
      path: nil,
      skip_multiline_encoding: nil,
      type: nil

    )

      response = run_command(create_secret: CreateSecretOptions.new(
        secret_name: secret_name,
        secret_value: secret_value,
        secret_comment: secret_comment,
        project_id: project_id,
        environment: environment,
        skip_multiline_encoding: skip_multiline_encoding,
        path: path,
        create_secret_options_type: type
      ))

      secrets_response = ResponseForCreateSecretResponse.from_json!(response).to_dynamic
      error_handler(secrets_response)


      secrets_response['data']['secret']
    end

    def delete(
      secret_name:,
      project_id:,
      environment:,
      path: nil,
      type: nil
    )
      response = run_command(delete_secret: DeleteSecretOptions.new(
        secret_name: secret_name,
        project_id: project_id,
        environment: environment,
        path: path,
        delete_secret_options_type: type
      ))

      secrets_response = ResponseForDeleteSecretResponse.from_json!(response).to_dynamic
      error_handler(secrets_response)

      secrets_response['data']['secret']
    end

    private

    def error_handler(response)

      # If the response is successful, we return without raising errors.
      if response.key?('success') && response['success'] == true && response.key?('data')
        return
      end

      if response['errorMessage']
        raise InfisicalError, response['errorMessage'] if response.key?('errorMessage')
      else
        raise InfisicalError, 'Error while getting response'
      end
    end

    def run_command(command)
      response = @command_runner.run(InfisicalCommands.new(command))
      raise InfisicalError, 'Error getting response' if response.nil?

      response
    end
  end
end

# rubocop:enable Metrics/ParameterLists
# rubocop:enable Metrics/MethodLength
